README.md
APP-SCIENCE-NGSpice
===================

Ngspice is a mixed-level/mixed-signal circuit simulator.

LICENSE
===============
* Several licenses

COMPILE TOOLS
===============
* 

AUTHORS
===============
* Port: Mentore Siesto

LINKS
===============
* http://ngspice.sourceforge.net/


ChangeLog
============================ Rework-22 ==================================
2010-10-04  Holger Vogt
 * miscoms.c: remove bug no. 3080398
 
2010-10-02  Holger Vogt
 * mesadefs.h, mesamsak.c, mesamparam.c, mesasetup.c
   jfetmask.c: remove NAN or exp -312 outputs

2010-10-01  Robert Larice
 * src/spicelib/devices/hfet1/hfetmask.c ,
 * src/spicelib/devices/hfet2/hfet2mask.c ,
 * src/spicelib/devices/mes/mesmask.c ,
 * src/spicelib/devices/mesa/mesamask.c :
    bugfix, missing return(OK) in some switch statements

2010-09-30  Robert Larice
 * configure.ac ,
 * src/include/defines.h :
    use autoconf to find a suitable signal function type

2010-09-30  Robert Larice
 * src/xspice/icm/Makefile :
    use $(MAKE) instead of `make'

2010-09-30  Robert Larice
 * src/xspice/cmpp/Makefile.am ,
 * src/xspice/cmpp/ifs_lex.l ,
 * src/xspice/cmpp/mod_lex.l :
    always use flex instead of an arbitrary lex,
    use flex specific %options, to clean up the Makefile
    don't write the lexer via stdout, to improve gcc warning readability
    this works around a bug as well, which caused erronous lexer c files
    to be included in the distribution tar ball

2010-09-26  Holger Vogt
 * .cvsignore: m4 entry removed

2010-09-25  Holger Vogt
 * conf.c, conf.h, ngspice.h, visualc/config.h, configure.ac:
   update welcome message, version 22
 * FAQ, COPYING, NEWS, INSTALL, README, README.tcl, BUGS: 
   update to version 22  
 * examples/Monte_Carlo: new folder with example files

2010-09-23  Robert Larice
 * Makefile.am ,
 * m4/.cvsignore :
    add an empty m4/.cvsignore file, to make sure the directory exists
    add this directory and the file compile_min.sh to the tarball

2010-09-22 Dietmar Warning
 * devices/mesa/mesaload.c: initialize some variables

2010-09-21  Robert Larice
 * autogen.sh :
    there are several implementations of `sed'.
    Some can't swallow a '\n' escape sequence on the RHS.
    And some want a plain newline in the RHS to be escaped.
    see http://sed.sourceforge.net/sedfaq4.html#s4.1
     `4.1. How do I insert a newline into the RHS of a substitution?'

2010-09-19 Dietmar Warning
 * adms/hicum2/admsva/hicum2.va: update to version 2.24 with small exceptions
   from the original source code, seems the last update because model provider
   has limited the access only to authorized partners

2010-09-19  Robert Larice
 * autogen.sh :
    cleanup

2010-09-19  Robert Larice
 * src/frontend/plotting/graf.c ,
 * src/frontend/plotting/grid.c :
    fix some ancient \n\r and \r\n typos

2010-09-19  Robert Larice
 * src/frontend/quote.c :
    cp_unqoute(), avoid segfault when applied to a single quote "\""
    testcase:
      .control
      echo "--"
      .endc
    FIXME, still not completely done,
      inp_stripcomments_line()
    will fail on comments inside of strings

2010-09-19 Holger Vogt
 * src/spinit.in, configure.ac: allow lib64 in spinit (bug no. 3054424)
   if libdir contains lib64
 * compile_min.sh: new script for compiling MINGW 32 and 64 bit

2010-09-18 Holger Vogt
 * src/XSPICE/icm/makedefs.in: external LDFLAGS enabled for MINGW

2010-09-18  Robert Larice
 * contrib/mslib/liblook ,
 * contrib/mslib/libprm ,
 * contrib/scripts/liblook ,
 * contrib/scripts/libprm ,
 * contrib/vbic/sgp2vbic ,
 * contrib/vbic/vbic2sgp *
    some scripts need bash
      this patch is from debian, thanks

2010-09-18 Dietmar Warning
 * src/spicelib/devices/*/isrcacct.c, vsrcacct.c: Correct bitwise operation

2010-09-17 Holger Vogt
 * visualc/include/config.h, vngspice.sln, vngspice.vcproj:
   new configurations release, Debug, and ReleaseOMP

2010-09-17  Robert Larice
 * .cvsignore ,
 * Makefile.am ,
 * configure.ac ,
 * config/.cvsignore :
    undo "use a subdirectory `config' for autoconfigure ..." from today
    this has broken `configure' for xgraph

2010-09-17  Robert Larice
 * src/frontend/Makefile.am :
    fix parse-bison.y dependencies

2010-09-17  Robert Larice
 * autogen.sh :
    run aclocal first, then libtoolize

2010-09-17  Robert Larice
 * Makefile.am ,
 * src/include/Makefile.am ,
 * src/maths/misc/Makefile.am ,
 * src/maths/poly/Makefile.am ,
 * src/spicelib/devices/ndev/Makefile.am ,
 * src/xspice/Makefile.am ,
 * tests/bsim3soi/Makefile.am ,
 * tests/bsim3soidd/Makefile.am ,
 * tests/bsim3soifd/Makefile.am ,
 * tests/bsim3soipd/Makefile.am ,
 * tests/mesa/Makefile.am :
    `make dist' add some missing artifacts

2010-09-17  Robert Larice
 * configure.ac ,
 * src/spicelib/devices/Makefile.am :
    `make dist' improvements for `NDEV'

2010-09-17  Robert Larice
 * Makefile.am ,
 * .cvsignore ,
 * config/.cvsignore ,
 * configure.ac :
    use a subdirectory `config' for autoconfigure related artifacts

2010-09-17  Robert Larice
 * **/Makefile.am :
    whitespace cleanup

2010-09-15 Holger Vogt
 * com_measure2.c: prevent seg fault in meas ac, if real input
 * examples/control_structs/MonteCarlo.sp: update, add measure

2010-09-12  Robert Larice
 * src/main.c :
    fix incorrect buffer type, and exit(1) on failure to open a tempfile

2010-09-08 Holger Vogt
 * inp.c, inpcom.c: include libiberty.h removed
 * terminal.c, ivars.c, ifnewuid.c, inperror.c, tclspice.c, main.c:
   rework of #ifdef HAVE_ASPRINTF ...
 * fteparse.h: #undef AND /* obsolete macro in ansidecl.h */ added

2010-09-08  Robert Larice
 * src/frontend/wdisp/windisp.c ,
 * src/frontend/wdisp/winprint.c :
    fix commit from today, `function rename, Arc() -> DevDrawArc()`

2010-09-08  Robert Larice
 * src/include/const.h :
    rename include file wrapper, #ifndef CONST  ->  CONST_H

2010-09-08  Robert Larice
 * src/frontend/display.c ,
 * src/include/ftedev.h :
    function rename, struct DISPDEVICE . Text() -> DrawText()

2010-09-08  Robert Larice
 * src/frontend/display.c ,
 * src/include/ftedev.h :
    function rename, struct DISPDEVICE . Arc() -> DrawArc()

2010-09-08  Robert Larice
 * src/frontend/display.c ,
 * src/frontend/display.h ,
 * src/frontend/plotting/graf.c ,
 * src/frontend/plotting/grid.c :

src/frontend/cpitf.c
/**********
Copyright 1990 Regents of the University of California.  All rights reserved.
Author: 1985 Wayne A. Christopher, U. C. Berkeley CAD Group
***********
$Id: cpitf.c,v 1.25 2010/09/08 17:14:44 rlar Exp $
*/


#include "ngspice.h"
#include "cpdefs.h"
#include "ftedefs.h"
#include "dvec.h"
#include "fteparse.h"
#include "cpitf.h"

#include <terminal.h>

#include "completion.h"
#include "variable.h"

/* Set some standard variables and aliases, etc, and init the ccom stuff.  
   Called by fcn main() */

void
ft_cpinit(void)
{
    wordlist *wl;
    wordlist wl1, wl2, wl3;
    bool found = FALSE, t = TRUE;
    char buf[BSIZE_SP], **x, *s, *r,*copys;
    struct comm *c;
    int i;
    FILE *fp;
    static char *predefs[] = {
        "yes",      "1",
        "TRUE",     "1",
        "no",       "0",
        "FALSE",    "0",
        "pi",       "3.14159265358979323846",
        "e",        "2.71828182844590452353",
        "c",        "2.997925e8",
        "i",        "0,1",
        "kelvin",   "-273.15",
        "echarge",  "1.60219e-19",
        "boltz",    "1.38062e-23",
        "planck",   "6.62620e-34"
    } ;
    static char *udfs[] = {
        "max(x,y)", "(x gt y) * x + (x le y) * y",
        "min(x,y)", "(x lt y) * x + (x ge y) * y",
        "vdb(x)",   "db(v(x))",
        "vdb(x,y)", "db(v(x) - v(y))",
        "vi(x)",    "im(v(x))",
        "vi(x,y)",  "im(v(x) - v(y))",
        "vm(x)",    "mag(v(x))",
        "vm(x,y)",  "mag(v(x) - v(y))",
        "vg(x)",    "group_delay(v(x))", //A.Rroldan 10/06/05 group delay new function
        "gd(x)",    "group_delay(v(x))", //A.Rroldan 10/06/05 group delay new function        
        "vp(x)",    "ph(v(x))",
        "vp(x,y)",  "ph(v(x) - v(y))",
        "vr(x)",    "re(v(x))",
        "vr(x,y)",  "re(v(x) - v(y))"
    } ;

    /* if TIOCSTI is defined (not available in MS Windows:
       Make escape the break character.
       So the user can type ahead...
       fcn defined in complete.c. */
    cp_ccon(TRUE);
    /* Initialize io, cp_chars[], variable "history" in init.c. */
    cp_init();

    /* If command completion is available (global variable cp_nocc
       set in main.c by command line option -q)   */ 
    if (!cp_nocc) {
        /* Add commands... */
        for (c = cp_coms; c->co_func; c++) {
            if (c->co_spiceonly && ft_nutmeg)
                continue;
            cp_addcomm(c->co_comname, c->co_cctypes[0],
                    c->co_cctypes[1], c->co_cctypes[2],
                    c->co_cctypes[3]);
            cp_addkword(CT_COMMANDS, c->co_comname);
        }
        /* And keywords... These are the ones that are constant... */
        if (!ft_nutmeg) {
            cp_addkword(CT_LISTINGARGS, "deck");
            cp_addkword(CT_LISTINGARGS, "logical");
            cp_addkword(CT_LISTINGARGS, "physical");
            cp_addkword(CT_LISTINGARGS, "expand");

            cp_addkword(CT_STOPARGS, "when");
            cp_addkword(CT_STOPARGS, "after");
        }

        cp_addkword(CT_PLOT, "new");

        cp_addkword(CT_PLOTKEYWORDS, "xlimit");
        cp_addkword(CT_PLOTKEYWORDS, "ylimit");
        cp_addkword(CT_PLOTKEYWORDS, "vs");
        cp_addkword(CT_PLOTKEYWORDS, "xindices");
        cp_addkword(CT_PLOTKEYWORDS, "xcompress");
        cp_addkword(CT_PLOTKEYWORDS, "xdelta");
        cp_addkword(CT_PLOTKEYWORDS, "ydelta");
        cp_addkword(CT_PLOTKEYWORDS, "lingrid");
        cp_addkword(CT_PLOTKEYWORDS, "loglog");
        cp_addkword(CT_PLOTKEYWORDS, "linear");
        cp_addkword(CT_PLOTKEYWORDS, "xlog");
        cp_addkword(CT_PLOTKEYWORDS, "ylog");
        cp_addkword(CT_PLOTKEYWORDS, "polar");
        cp_addkword(CT_PLOTKEYWORDS, "smith");
        cp_addkword(CT_PLOTKEYWORDS, "smithgrid");
        cp_addkword(CT_PLOTKEYWORDS, "nointerp");
        cp_addkword(CT_PLOTKEYWORDS, "title");
        cp_addkword(CT_PLOTKEYWORDS, "xlabel");
        cp_addkword(CT_PLOTKEYWORDS, "ylabel");
        cp_addkword(CT_PLOTKEYWORDS, "linplot");
        cp_addkword(CT_PLOTKEYWORDS, "combplot");
        cp_addkword(CT_PLOTKEYWORDS, "pointplot");

        cp_addkword(CT_RUSEARGS, "time");
        cp_addkword(CT_RUSEARGS, "space");
        cp_addkword(CT_RUSEARGS, "faults");
        cp_addkword(CT_RUSEARGS, "elapsed");
        cp_addkword(CT_RUSEARGS, "totiter");
        cp_addkword(CT_RUSEARGS, "traniter");
        cp_addkword(CT_RUSEARGS, "tranpoints");
        cp_addkword(CT_RUSEARGS, "accept");
        cp_addkword(CT_RUSEARGS, "rejected");
        cp_addkword(CT_RUSEARGS, "time");
        cp_addkword(CT_RUSEARGS, "trantime");
        cp_addkword(CT_RUSEARGS, "lutime");
        cp_addkword(CT_RUSEARGS, "solvetime");
        cp_addkword(CT_RUSEARGS, "transolvetime");
        cp_addkword(CT_RUSEARGS, "loadtime");
        cp_addkword(CT_RUSEARGS, "all");

        cp_addkword(CT_VECTOR, "all");

        for (x = ft_setkwords; *x; x++)
            cp_addkword(CT_VARIABLES, *x);
        for (i = 0; ; i++) {
            if (!(s = ft_typenames(i)))
                break;
            cp_addkword(CT_TYPENAMES, s);
        }
    }

    cp_vset("program", CP_STRING, cp_program);

    /* Make the prompt use only the last component of the path... */

    if (DIR_TERM) {
	for (s = cp_program; s && *s; s++)
	    ;
	s--;
	while ((s > cp_program) && (*s != DIR_TERM))
	    s--;
	if (*s == DIR_TERM)
	    s++;
	(void) strcpy(buf, s);
	for (s = buf; *s && (*s != '.'); s++)
	    ;
	*s = '\0';
	(void) strcat(buf, " ! -> ");
    } else
	(void) sprintf(buf, "%s ! -> ", cp_program);

    cp_vset("prompt", CP_STRING, buf);
    cp_vset("noglob", CP_BOOL, &t);
    cp_vset("brief", CP_BOOL, &t);

    /* Make vectors from values in predefs[] for the current plot. 
     Define functions from entries in udfs[] (like user defined functions).
     */
    wl1.wl_next = &wl2;
    wl1.wl_prev = NULL;
    wl2.wl_next = NULL;
    wl2.wl_prev = &wl1;
    wl1.wl_word = "if";
    wl2.wl_word = "1";
    cp_setalias("begin", &wl1);
    wl1.wl_next = NULL;
    wl1.wl_word = "end";
    cp_setalias("endif", &wl1);
    cp_setalias("endwhile", &wl1);
    cp_setalias("endforeach", &wl1);
    cp_setalias("endrepeat", &wl1);
    cp_setalias("enddowhile", &wl1);
    wl1.wl_word = "help";
    cp_setalias("?", &wl1);

    wl1.wl_next = &wl2;
    wl2.wl_next = &wl3;
    wl2.wl_prev = &wl1;
    wl3.wl_prev = &wl2;
    wl3.wl_next = NULL;
    wl2.wl_word = "=";
    for (i = 0; i < sizeof (predefs) / sizeof (char *); i += 2) {
        wl1.wl_word = predefs[i];
        wl3.wl_word = predefs[i + 1];
        com_let(&wl1);
    }

    wl2.wl_next = NULL;
    for (i = 0; i < sizeof (udfs) / sizeof (char *); i += 2) {
        wl1.wl_word = udfs[i];
        wl2.wl_word = udfs[i + 1];
        com_define(&wl1);
    }

    /* Reset this for the front end. */
    cp_hash = '*';

    /* NGSPICEDATADIR has been set to path "$dprefix/share/ngspice" in configure.ac, 
       Spice_Lib_Dir has been set to NGSPICEDATADIR in conf.c, 
       may be overridden by environmental variable SPICE_LIB_DIR in ivars().
       Lib_Path has been set to Spice_Lib_Dir adding /scripts in ivars() */
    if (Lib_Path && *Lib_Path) {
       /* set variable 'sourcepath' */ 
       (void) sprintf(buf, "sourcepath = ( %s %s )", DIR_CWD, Lib_Path);
        wl = cp_doglob(cp_lexer(buf));
        cp_striplist(wl);
        com_set(wl);
        wl_free(wl);
        
        /* Now source the standard startup file spinit or tclspinit. */

        /* jump over leading spaces */
        for (copys=s=cp_tildexpand(Lib_Path); copys && *copys; ) {
            while (isspace(*s))
                s++;
            /* copy s into buf until space is seen, r is the actual position */
            for (r = buf; *s && !isspace(*s); r++, s++)
                *r = *s;
            tfree(copys);
	         /* add a path separator to buf at actual position */
            (void) strcpy(r, DIR_PATHSEP);
#ifdef TCL_MODULE
            /* add "tclspinit" to buf after actual position */
            (void) strcat(r, "tclspinit");
#else
            /* add "spinit" to buf after actual position */
            (void) strcat(r, "spinit");
#endif
            if ((fp = fopen(buf, "r"))) {
                cp_interactive = FALSE;
                inp_spsource(fp, TRUE, buf);
                cp_interactive = TRUE;
                found = TRUE;
                break;
#if defined (HAS_WINDOWS) || defined (__MINGW32__) || defined (_MSC_VER)
            /* search in local directory where ngspice.exe resides */
            } else if ((fp = fopen("./spinit", "r"))) {
                cp_interactive = FALSE;
                inp_spsource(fp, TRUE, buf);
                cp_interactive = TRUE;
                found = TRUE;
                break;
#endif                
            } else if (ft_controldb)
                fprintf(cp_err, "Note: can't open \"%s\".\n", buf);
        }
        if (!found)
            fprintf(cp_err, "Note: can't find init file.\n");
    }

    tcap_init( );
    return;
}

/* Decide whether a condition is TRUE or not. */

bool
cp_istrue(wordlist *wl)
{
    int i;
    struct dvec *v;
    struct pnode *pn;

/* fprintf(stderr, "isTRUE: "); wl_print(wl, stderr); fprintf(stderr, "\n"); */
    /* First do all the csh-type stuff here... */
    wl = wl_copy(wl);
    wl = cp_variablesubst(wl);
    wl = cp_bquote(wl);
    cp_striplist(wl);

    pn = ft_getpnames(wl, TRUE);
    wl_free(wl);
    v = ft_evaluate(pn);

    /* It makes no sense to say while (all), but what the heck... */
    while (v) {
        if (isreal(v)) {
            for (i = 0; i < v->v_length; i++)
                if (v->v_realdata[i] != 0.0) {
		   free_pnode(pn);
                   return (TRUE);
		}
        } else {
            for (i = 0; i < v->v_length; i++)
                if ((realpart(&v->v_compdata[i]) != 0.0) ||
                    (imagpart(&v->v_compdata[i]) != 0.0)) {
		    free_pnode(pn);
                    return (TRUE);
		}
        }
        v = v->v_link2;
    }
   free_pnode(pn);
   return (FALSE);
}

/* This gets called before every command is executed... 
   from fcns do_command() or do_block() in control.c */

void
cp_periodic(void)
{
    ft_setflag = FALSE;
    ft_intrpt = FALSE;
    ft_ckspace();   /* check for max. data size (resource.c) */
    ft_checkkids(); /* check for jobs (only if OK_ASPICE is defined, apsice.c) */
    vec_gc();       /* remove vectors which do not have permanent flag set (vectors.c) */
    return;
}

void
cp_doquit(void)
{
    com_quit(NULL);
    return;
}

/* This is how we deal with emulation of commands by scripts... If the script
 * is found, then set the variables argc and argv and call the script.  Note
 * that this also allows us to type a filename to load a spice deck...
 */

bool
cp_oddcomm(char *s, wordlist *wl)
{
    FILE *fp;
    char buf[BSIZE_SP];
    wordlist ww;

    if ((fp = inp_pathopen(s, "r"))) {
        (void) fclose(fp);
        (void) sprintf(buf, "argc = %d argv = ( ", wl_length(wl));
        while (wl) {
            (void) strcat(buf, wl->wl_word);
            (void) strcat(buf, " ");
            wl = wl->wl_next;
        }
        (void) strcat(buf, ")");
        com_set(cp_lexer(buf));
        inp_source(s);
        cp_remvar("argc");
        cp_remvar("argv");
        return (TRUE);
    } else if (wl && eq(wl->wl_word, "=")) {
        ww.wl_next = wl;
        wl->wl_prev = &ww;
        ww.wl_prev = NULL;
        ww.wl_word = s;
        com_let(&ww);
        return (TRUE);
    } else
        return (FALSE);
}

